File: common/libs/ConfigManager/python/etcd/etcd_client.py
# Copyright (c) 2019 Intel Corporation.

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
etcd_client module is a etc wrapper to read any keys and watch on a directory
in etcd distributed key store
"""

import etcd3
import logging
import json
import os
from util.util import Util


class EtcdCli:

    def __init__(self, config):
        """ constructor which creates an EtcdCli instance, checks for
        the etcd service port availability
        :param config: config of type Dict with certFile, keyFile
        and trustFile"""

        self.logger = logging.getLogger(__name__)
        hostname = "localhost"

        # This change will be moved to an argument to the function in 2.3
        # This is done now for backward compatibility
        etcd_host = os.getenv("ETCD_HOST")
        if etcd_host is not None and etcd_host != "":
            hostname = etcd_host

        port = 2379

        if not Util.check_port_availability(hostname, port):
            raise Exception("etcd service port {} is not up!".format(port))

        try:
            if config["trustFile"] == "" and config["keyFile"] == "" \
               and config["certFile"] == "":
                self.etcd = etcd3.client(host=hostname, port=port)
            else:
                self.etcd = etcd3.client(host=hostname, port=port,
                                         ca_cert=config["trustFile"],
                                         cert_key=config["keyFile"],
                                         cert_cert=config["certFile"])
        except Exception as e:
            raise e
        self.callback = None
        self._setEnv()

    def _setEnv(self):
        """ _setEnv is a local function to set global env """
        value = self.etcd.get("/GlobalEnv/")
        jsonConfig = json.loads(value[0].decode('utf-8'))
        for key in jsonConfig.keys():
            os.environ[key] = jsonConfig[key]

    def GetConfig(self, key):
        """ GetConfig gets the value of a key from Etcd
        :param key: key to be queried on, form etcd
        :type: string
        :return: values returned from etcd based on key
        :rtype: string"""
        try:
            value = self.etcd.get(key)
        except Exception as e:
            raise e
        if value[0] is not None:
            return value[0].decode('utf-8')
        return value[0]

    def PutConfig(self, key, value):
        """ PutConfig to save a value to Etcd
        :param key: keyin etcd to set
        :type: string
        :param value: value to set key to
        :type: string
        """
        try:
            self.etcd.put(key, value)
        except Exception as e:
            raise e

    def onChangeCallback(self, event):
        key = event.events[0].key
        value = event.events[0].value
        self.callback(key.decode('utf-8'), value.decode('utf-8'))

    def RegisterDirWatch(self, key, callback):
        """ RegisterDirWatch registers to a callback and keeps a watch on the
        prefix of a specified key
        :param key: prefix of a key to keep a watch on, in etcd
        :type string
        :callback : callback function
        :return None"""

        self.callback = callback
        try:
            self.etcd.add_watch_prefix_callback(key, self.onChangeCallback)
        except Exception as e:
            raise e

    def RegisterKeyWatch(self, key, callback):
        """ RegisterKeyWatch registers to a callback and keeps a watch
        on a specified key
        :param key: key to keep a watch on, in etcd
        :type string
        :callback : callback function
        :return None"""

        self.callback = callback
        try:
            self.etcd.add_watch_callback(key, self.onChangeCallback)
        except Exception as e:
            raise e

File: common/libs/ConfigManager/go/EtcdClient.go
/*
Copyright (c) 2019 Intel Corporation.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

// Package configmanager is a configmanager wrapper to read any keys and watch on a directory in etcd distributed key store
package configmanager

import (
	util "IEdgeInsights/common/util"
	"context"
	"encoding/json"
	"errors"
	"os"

	"github.com/golang/glog"
	"go.etcd.io/etcd/clientv3"
	"go.etcd.io/etcd/pkg/transport"
)

// EtcdCli implements ConfigMgr
type EtcdCli struct {
	etcd *clientv3.Client
}

// RegisterWatchOnKey go routine
func RegisterWatchOnKey(rch clientv3.WatchChan, onChangeCallback OnChangeCallback) {

	for wresp := range rch {
		for _, ev := range wresp.Events {
			onChangeCallback(convertUIntArrToString(ev.Kv.Key), convertUIntArrToString(ev.Kv.Value))
		}
	}

}

// NewEtcdClient constructs a new EtcdClient and checks etcd port's availability
func NewEtcdClient(conf config) (etcdCli *EtcdCli, err error) {

	var cfg clientv3.Config
	hostname := "localhost"

	// This change will be moved to an argument to the function in 2.3
	// This is done now for backward compatibility
	etcdHost := os.Getenv("ETCD_HOST")
	if etcdHost != "" {
		hostname = etcdHost
	}

	port := "2379"
	endpoint := []string{hostname + ":" + port}
	portUp := util.CheckPortAvailability(hostname, port)
	if !portUp {
		glog.Error("etcd service port not up")
		return nil, errors.New("etcd service port not up")
	}

	devmode := false
	if conf.certFile == "" && conf.keyFile == "" && conf.trustFile == "" {
		devmode = true
	}

	if devmode {
		cfg = clientv3.Config{
			Endpoints: endpoint,
		}
	} else {
		tlsInfo := transport.TLSInfo{
			CertFile:      conf.certFile,
			KeyFile:       conf.keyFile,
			TrustedCAFile: conf.trustFile,
		}

		tlsConfig, err := tlsInfo.ClientConfig()
		if err != nil {
			return nil, err
		}

		cfg = clientv3.Config{
			Endpoints: endpoint,
			TLS:       tlsConfig,
		}
	}

	etcdcli, err := clientv3.New(cfg)
	if err != nil {
		return nil, err
	}
	_setEnv(etcdcli)
	return &EtcdCli{etcd: etcdcli}, err
}

// _setEnv is a local function to set global env
func _setEnv(etcd *clientv3.Client) {
	response, err := etcd.Get(context.TODO(), "/GlobalEnv/")
	if err != nil {
		return
	}

	responseString := ""
	for _, ev := range response.Kvs {
		responseString = convertUIntArrToString(ev.Value)
	}

	// Setting env variables
	jsonMap := make(map[string]interface{})
	json.Unmarshal([]byte(responseString), &jsonMap)
	for key, value := range jsonMap {
		os.Setenv(key, value.(string))
	}
}

// GetConfig gets the value of a key from Etcd
func (etcdClient EtcdCli) GetConfig(key string) (string, error) {

	response, err := etcdClient.etcd.Get(context.TODO(), key)
	if err != nil {
		return "", err
	}
	responseString := ""
	for _, ev := range response.Kvs {
		responseString = convertUIntArrToString(ev.Value)
	}
	return responseString, err
}

// PutConfig to Save a value of the key to ETCD
func (etcdClient EtcdCli) PutConfig(key string, value string) error {
	_, err := etcdClient.etcd.Put(context.TODO(), key, value)
	if err != nil {
		return err
	}
	return nil
}

// RegisterDirWatch registers to a callback and keeps a watch on the prefix of a specified key
func (etcdClient *EtcdCli) RegisterDirWatch(key string, onChangeCallback OnChangeCallback) {

	glog.Infoln("watching on Dir:", key)
	rch := etcdClient.etcd.Watch(context.TODO(), key, clientv3.WithPrefix())
	go RegisterWatchOnKey(rch, onChangeCallback)

}

// RegisterKeyWatch registers to a callback and keeps a watch on a specified key
func (etcdClient *EtcdCli) RegisterKeyWatch(key string, onChangeCallback OnChangeCallback) {

	glog.Infoln("Watching on key:", key)
	rch := etcdClient.etcd.Watch(context.TODO(), key)
	go RegisterWatchOnKey(rch, onChangeCallback)

}

//convertUIntArrToString function to convert unit8[] to string
func convertUIntArrToString(uintarray []uint8) string {
	b := make([]byte, len(uintarray))
	for i, v := range uintarray {
		b[i] = byte(v)
	}
	return string(b)
}
